File: src/main/java/pl/mjedynak/idea/plugins/builder/action/handler/DisplayChoosers.java
package pl.mjedynak.idea.plugins.builder.action.handler;

import com.intellij.codeInsight.generation.PsiElementClassMember;
import com.intellij.ide.util.MemberChooser;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiClass;
import com.intellij.psi.PsiDirectory;
import com.intellij.psi.PsiFile;
import java.util.List;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import pl.mjedynak.idea.plugins.builder.factory.CreateBuilderDialogFactory;
import pl.mjedynak.idea.plugins.builder.factory.MemberChooserDialogFactory;
import pl.mjedynak.idea.plugins.builder.factory.PsiFieldsForBuilderFactory;
import pl.mjedynak.idea.plugins.builder.gui.CreateBuilderDialog;
import pl.mjedynak.idea.plugins.builder.psi.PsiFieldSelector;
import pl.mjedynak.idea.plugins.builder.psi.PsiHelper;
import pl.mjedynak.idea.plugins.builder.psi.model.PsiFieldsForBuilder;
import pl.mjedynak.idea.plugins.builder.writer.BuilderContext;
import pl.mjedynak.idea.plugins.builder.writer.BuilderWriter;

public class DisplayChoosers {

    private final @NotNull PsiClass psiClassFromEditor;
    private final @NotNull Project project;
    private final @NotNull Editor editor;

    public DisplayChoosers(@NotNull PsiClass psiClassFromEditor, @NotNull Project project, @NotNull Editor editor) {
        this.psiClassFromEditor = psiClassFromEditor;
        this.project = project;
        this.editor = editor;
    }

    public void run(@Nullable PsiClass existingBuilder) {
        CreateBuilderDialog createBuilderDialog = showDialog(existingBuilder);
        if (createBuilderDialog != null && createBuilderDialog.isOK()) {
            PsiDirectory targetDirectory = createBuilderDialog.getTargetDirectory();
            String className = createBuilderDialog.getClassName();
            String methodPrefix = createBuilderDialog.getMethodPrefix();
            boolean innerBuilder = createBuilderDialog.isInnerBuilder();
            boolean useSingleField = createBuilderDialog.useSingleField();
            boolean hasButMethod = createBuilderDialog.hasButMethod();
            List<PsiElementClassMember<?>> fieldsToDisplay =
                    getFieldsToIncludeInBuilder(psiClassFromEditor, innerBuilder, useSingleField, hasButMethod);
            MemberChooser<PsiElementClassMember<?>> memberChooserDialog =
                    MemberChooserDialogFactory.getMemberChooserDialog(fieldsToDisplay, project);
            memberChooserDialog.show();
            writeBuilderIfNecessary(
                    targetDirectory,
                    className,
                    methodPrefix,
                    memberChooserDialog,
                    createBuilderDialog,
                    existingBuilder);
        }
    }

    private void writeBuilderIfNecessary(
            PsiDirectory targetDirectory,
            String className,
            String methodPrefix,
            @NotNull MemberChooser<PsiElementClassMember<?>> memberChooserDialog,
            CreateBuilderDialog createBuilderDialog,
            @Nullable PsiClass existingBuilder) {
        if (memberChooserDialog.isOK()) {
            List<PsiElementClassMember<?>> selectedElements = memberChooserDialog.getSelectedElements();
            if (selectedElements == null) return;
            PsiFieldsForBuilder psiFieldsForBuilder =
                    PsiFieldsForBuilderFactory.createPsiFieldsForBuilder(selectedElements, psiClassFromEditor);
            BuilderContext context = new BuilderContext(
                    project,
                    psiFieldsForBuilder,
                    targetDirectory,
                    className,
                    psiClassFromEditor,
                    methodPrefix,
                    createBuilderDialog.isInnerBuilder(),
                    createBuilderDialog.hasButMethod(),
                    createBuilderDialog.useSingleField(),
                    createBuilderDialog.hasAddCopyConstructor());
            BuilderWriter.writeBuilder(context, existingBuilder);
        }
    }

    private @Nullable CreateBuilderDialog showDialog(@Nullable PsiClass existingBuilder) {
        PsiFile file = PsiHelper.getPsiFile(editor, project);
        if (file == null) return null;
        CreateBuilderDialog dialog = CreateBuilderDialogFactory.createBuilderDialog(
                psiClassFromEditor, project, PsiHelper.getPackage(file.getContainingDirectory()), existingBuilder);
        dialog.show();
        return dialog;
    }

    private static @NotNull List<PsiElementClassMember<?>> getFieldsToIncludeInBuilder(
            PsiClass clazz, boolean innerBuilder, boolean useSingleField, boolean hasButMethod) {
        return PsiFieldSelector.selectFieldsToIncludeInBuilder(clazz, innerBuilder, useSingleField, hasButMethod);
    }
}

File: src/main/kotlin/io/github/kingg22/kobra/builder/action/handler/DisplayChoosers.kt
package io.github.kingg22.kobra.builder.action.handler

import com.intellij.codeInsight.generation.PsiElementClassMember
import com.intellij.ide.util.MemberChooser
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiClass
import com.intellij.psi.PsiDirectory
import com.intellij.psi.PsiFile
import io.github.kingg22.kobra.builder.factory.CreateBuilderDialogFactory
import io.github.kingg22.kobra.builder.factory.MemberChooserDialogFactory
import io.github.kingg22.kobra.builder.factory.PsiFieldsForBuilderFactory
import io.github.kingg22.kobra.builder.gui.CreateBuilderDialog
import io.github.kingg22.kobra.builder.psi.PsiFieldSelector
import io.github.kingg22.kobra.builder.psi.PsiHelper
import io.github.kingg22.kobra.builder.writer.BuilderContext
import io.github.kingg22.kobra.builder.writer.BuilderWriter

public class DisplayChoosers(
    private val psiClassFromEditor: PsiClass,
    private val project: Project,
    private val editor: Editor,
) {
    public fun run(existingBuilder: PsiClass?) {
        val createBuilderDialog = showDialog(existingBuilder)
        if (createBuilderDialog != null && createBuilderDialog.isOK) {
            val targetDirectory = createBuilderDialog.targetDirectory
            val className = createBuilderDialog.className.orEmpty()
            val methodPrefix = createBuilderDialog.methodPrefix
            val innerBuilder = createBuilderDialog.isInnerBuilder
            val useSingleField = createBuilderDialog.isSingleField
            val hasButMethod = createBuilderDialog.hasButMethod
            val fieldsToDisplay =
                getFieldsToIncludeInBuilder(psiClassFromEditor, innerBuilder, useSingleField, hasButMethod)
            val memberChooserDialog = MemberChooserDialogFactory.getMemberChooserDialog(fieldsToDisplay, project)
            memberChooserDialog.show()
            writeBuilderIfNecessary(
                targetDirectory,
                className,
                methodPrefix,
                memberChooserDialog,
                createBuilderDialog,
                existingBuilder,
            )
        }
    }

    private fun writeBuilderIfNecessary(
        targetDirectory: PsiDirectory?,
        className: String,
        methodPrefix: String?,
        memberChooserDialog: MemberChooser<PsiElementClassMember<*>>,
        createBuilderDialog: CreateBuilderDialog,
        existingBuilder: PsiClass?,
    ) {
        if (memberChooserDialog.isOK) {
            val selectedElements = memberChooserDialog.getSelectedElements() ?: return
            val psiFieldsForBuilder = PsiFieldsForBuilderFactory.createPsiFieldsForBuilder(
                selectedElements,
                psiClassFromEditor,
            )
            val context = BuilderContext(
                project,
                psiFieldsForBuilder,
                targetDirectory,
                className,
                psiClassFromEditor,
                methodPrefix,
                createBuilderDialog.isInnerBuilder,
                createBuilderDialog.hasButMethod,
                createBuilderDialog.isSingleField,
                createBuilderDialog.hasAddCopyConstructor,
            )
            BuilderWriter.writeBuilder(context, existingBuilder)
        }
    }

    private fun showDialog(existingBuilder: PsiClass?): CreateBuilderDialog? {
        val file: PsiFile = PsiHelper.getPsiFile(editor, project) ?: return null
        val dialog = CreateBuilderDialogFactory.createBuilderDialog(
            psiClassFromEditor,
            project,
            PsiHelper.getPackage(file.containingDirectory),
            existingBuilder,
        )
        dialog.show()
        return dialog
    }

    public companion object {
        private fun getFieldsToIncludeInBuilder(
            clazz: PsiClass,
            innerBuilder: Boolean,
            useSingleField: Boolean,
            hasButMethod: Boolean,
        ): List<PsiElementClassMember<*>> = PsiFieldSelector.selectFieldsToIncludeInBuilder(
            clazz,
            innerBuilder,
            useSingleField,
            hasButMethod,
        )
    }
}
